### PR TITLE
Introduces citus_stat_activity view

### PR DESCRIPTION
**Pull request by halilozanakgul**
DESCRIPTION: Introduces citus_stat_activity view

This PR introduces `citus_stat_activity` view that combines `pg_stat_activity` views from all nodes.
 `citus_stat_activity` includes all `pg_stat_activity` columns from PG12, PG13, and PG14. PG13 and PG14 both added one more column to `pg_stat_activity`. `citus_stat_activity` includes both of them. If you're using a PG version where `citus_stat_activity` has a column `pg_stat_activity` doesn't, the value for that column will be `NULL`. And `citus_stat_activity` has three extra columns:`global_pid`, `nodeid` and `is_worker_query`.

This PR also introduces `run_command_on_all_nodes` UDF. This UDF is similar to `run_command_on_workers` but runs the command on coordinator too in addition to workers.
Only exception is that if the coordinator is not added to metadata and this command is called from a worker node, it will not be able to run the command on the coordinator.

### CODE REVIEW

File: src/backend/distributed/sql/udfs/citus_stat_activity/11.0-1.sql [hunk 1]
```diff
@@ -0,0 +1,13 @@
+DROP VIEW IF EXISTS pg_catalog.citus_stat_activity;
+
+CREATE VIEW citus.citus_stat_activity AS
+SELECT global_pid, nodeid, pg_stat_activity_from_all_nodes.* FROM
+run_command_on_workers($$
+    SELECT to_json(array_agg(pg_stat_activity.*)) FROM pg_stat_activity
+$$) as text_result_from_all_nodes
```
Comment (onderkalaci) [inline src/backend/distributed/sql/udfs/citus_stat_activity/11.0-1.sql:7]: this should include the coalesce discussed here: https://github.com/citusdata/citus/issues/3172#issuecomment-1047562173

Otherwise the query fails if any node returns zero rows. With pg_stat_activity that cannot happen, but still

File: src/backend/distributed/sql/udfs/citus_stat_activity/latest.sql [hunk 2]
```diff
@@ -0,0 +1,13 @@
+DROP VIEW IF EXISTS pg_catalog.citus_stat_activity;
+
+CREATE VIEW citus.citus_stat_activity AS
+SELECT global_pid, nodeid, pg_stat_activity_from_all_nodes.* FROM
+run_command_on_workers($$
+    SELECT to_json(array_agg(pg_stat_activity.*)) FROM pg_stat_activity
+$$) as text_result_from_all_nodes
+JOIN LATERAL json_populate_recordset(NULL::pg_stat_activity, text_result_from_all_nodes.result::JSON) as pg_stat_activity_from_all_nodes ON TRUE
+LEFT JOIN pg_dist_node ON text_result_from_all_nodes.nodename = pg_dist_node.nodename AND text_result_from_all_nodes.nodeport = pg_dist_node.nodeport
+LEFT JOIN get_global_active_transactions() ON pid = process_id AND initiator_node_identifier = pg_dist_node.nodeid;
```
Comment (onderkalaci) [inline src/backend/distributed/sql/udfs/citus_stat_activity/latest.sql:10]: hmm, do we set `initiator_node_identifier` properly in get_global_active_transactions?

It is currently set: `		initiatorNodeIdentifier = currentBackend->citusBackend.initiatorNodeIdentifier;`


But, we should extract it from gpid?

Comment (halilozanakgul) [reply src/backend/distributed/sql/udfs/citus_stat_activity/latest.sql:10]: Then let me create the `extract_node_id_from_gpid` udf

File: src/backend/distributed/sql/udfs/citus_stat_activity/latest.sql [hunk 3]
```diff
@@ -0,0 +1,13 @@
+DROP VIEW IF EXISTS pg_catalog.citus_stat_activity;
+
+CREATE VIEW citus.citus_stat_activity AS
+SELECT global_pid, nodeid, pg_stat_activity_from_all_nodes.* FROM
+run_command_on_workers($$
```
Comment (onderkalaci) [inline src/backend/distributed/sql/udfs/citus_stat_activity/latest.sql:5]: this should show the commands on the coordinator as well

File: src/backend/distributed/sql/udfs/citus_stat_activity/latest.sql (hunk 2 above)
Comment (halilozanakgul) [reply src/backend/distributed/sql/udfs/citus_stat_activity/latest.sql]: This turned out to be an incorrect approach. Because we are interested in the node id of the node the query runs on, not the initiator. But both `initiator_node_indentifier` and `extract_node_id_from_global_pid(global_pid)` are initiator node ids.

File: src/backend/distributed/sql/udfs/citus_stat_activity/latest.sql [hunk 4]
```diff
@@ -0,0 +1,13 @@
+DROP VIEW IF EXISTS pg_catalog.citus_stat_activity;
+
+CREATE VIEW citus.citus_stat_activity AS
+SELECT global_pid, nodeid, pg_stat_activity_from_all_nodes.* FROM
+run_command_on_all_nodes($$
```
Comment (onderkalaci) [inline src/backend/distributed/sql/udfs/citus_stat_activity/latest.sql:5]: maybe we should use `parallel=true`? what is your take?

File: src/backend/distributed/sql/udfs/run_command_on_all_nodes/11.0-1.sql [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+	RETURNS SETOF record
+	LANGUAGE plpgsql
+	AS $function$
+DECLARE
+	workers text[];
+	ports int[];
+	commands text[];
+BEGIN
+    WITH citus_nodes AS (
+        SELECT * FROM pg_dist_node ORDER BY nodename, nodeport
```
Comment (onderkalaci) [inline src/backend/distributed/sql/udfs/run_command_on_all_nodes/11.0-1.sql:13]: My comment somehow doesn't show up. We should filter with nodecluster  and maybe even `isActive`

File: src/backend/distributed/sql/udfs/citus_stat_activity/latest.sql (hunk 4 above)
Comment (halilozanakgul) [reply src/backend/distributed/sql/udfs/citus_stat_activity/latest.sql]: I guess that is OK and it is also the default.

File: src/backend/distributed/sql/udfs/citus_stat_activity/latest.sql [hunk 6]
```diff
@@ -0,0 +1,23 @@
+DROP VIEW IF EXISTS pg_catalog.citus_stat_activity;
+DROP TYPE IF EXISTS pg_catalog.global_pid_and_pg_stat_activity;
+
+CREATE TYPE citus.global_pid_and_pg_stat_activity AS (global_pid bigint, pg_stat_activity pg_stat_activity);
```
Comment (onderkalaci) [inline src/backend/distributed/sql/udfs/citus_stat_activity/latest.sql:4]: What happens to `global_pid_and_pg_stat_activity` whe `pg_stat_activity` changes in the next PG version?

File: src/backend/distributed/sql/udfs/citus_stat_activity/latest.sql [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+    SELECT * FROM run_command_on_all_nodes($$
+        SELECT coalesce(to_json(array_agg(gpid_and_psa.*)), '[{}]'::JSON)
+        FROM (
+            SELECT global_pid, (pg_stat_activity.*)::pg_stat_activity FROM
+            pg_stat_activity LEFT JOIN get_all_active_transactions() ON process_id = pid
+        ) AS gpid_and_psa;
+    $$)
+    WHERE success = 't'
+) AS run_command_on_all_nodes
+LEFT JOIN LATERAL json_populate_recordset(NULL::global_pid_and_pg_stat_activity, run_command_on_all_nodes.result::JSON) as gpid_and_psa_from_all_nodes ON TRUE;
```
Comment (onderkalaci) [inline src/backend/distributed/sql/udfs/citus_stat_activity/latest.sql:21]: Sorry, I misguided you in the chat. I think this join should not be a LEFT JOIN, that seems redundant

File: src/backend/distributed/sql/udfs/citus_stat_activity/latest.sql (hunk 6 above)
Comment (halilozanakgul) [reply src/backend/distributed/sql/udfs/citus_stat_activity/latest.sql:4]: We drop it and re-create it

File: src/test/regress/expected/isolation_global_pid.out [hunk 8]
```diff
@@ -30,17 +30,29 @@ run_commands_on_session_level_connection_to_node
 
 (1 row)
 
-step s2-coordinator-citus_dist_stat_activity:
-    SELECT global_pid != 0 FROM citus_dist_stat_activity() WHERE query LIKE '%SELECT * FROM dist\_table%' and query NOT ILIKE '%run_commands_on_session_level_connection_to_node%';
+step s2-coordinator-citus_stat_activity:
+    SELECT global_pid != 0 FROM citus_stat_activity WHERE query LIKE '%SELECT * FROM dist\_table%' AND query NOT ILIKE '%run_commands_on_session_level_connection_to_node%';
```
Comment (onderkalaci) [inline src/test/regress/expected/isolation_global_pid.out:34]: maybe this PR is a good one to remove all citus_dist/worker_stat_activity  from the tests? Or, do you want to do it in another PR?

File: src/backend/distributed/sql/udfs/citus_stat_activity/latest.sql (hunk 6 above)
Comment (onderkalaci) [reply src/backend/distributed/sql/udfs/citus_stat_activity/latest.sql:4]: I meant, PG version change. We execute these scripts on Citus version upgrades.

File: src/backend/distributed/sql/udfs/run_command_on_all_nodes/latest.sql [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+	RETURNS SETOF record
+	LANGUAGE plpgsql
+	AS $function$
+DECLARE
+	nodenames text[];
+	ports int[];
+	commands text[];
+BEGIN
+	WITH citus_nodes AS (
+		SELECT * FROM pg_dist_node WHERE isactive = 't' ORDER BY nodename, nodeport
```
Comment (marcocitus) [inline src/backend/distributed/sql/udfs/run_command_on_all_nodes/latest.sql:13]: should exclude nodes from different clusters (read replicas)

File: src/test/regress/expected/isolation_global_pid.out (hunk 8 above)
Comment (halilozanakgul) [reply src/test/regress/expected/isolation_global_pid.out:34]: There are some more tests so I'd prefer a separate PR.

File: src/backend/distributed/sql/udfs/run_command_on_all_nodes/latest.sql (hunk 9 above)
Comment (onderkalaci) [reply src/backend/distributed/sql/udfs/run_command_on_all_nodes/latest.sql:13]: (I wrote this comment 3rd time now, as Ozan pushes new changes it somehow gets lost:) )

Also probably filter inactive nodes as well because those might be inaccessible.


@halilozanakgul what happens if one of the nodes cannot be reached? `citus_dist_stat_activity` gives a warning but still returns the result from healthy nodes. I think that behavior is better (over throwing an error and not showing anything). Is it similar for this one?

Comment (halilozanakgul) [reply src/backend/distributed/sql/udfs/run_command_on_all_nodes/latest.sql:13]: I saw the 2nd one 😄 I'm just thinking how to do it.

> @halilozanakgul what happens if one of the nodes cannot be reached?

`citus_stat_activity` filters with the success column from this function. So only the results of successful commands but no warning.

Comment (onderkalaci) [reply src/backend/distributed/sql/udfs/run_command_on_all_nodes/latest.sql:13]: > citus_stat_activity filters with the success column from this function. So only the results of successful commands but no warning.

How hard is it to print the warnings? Wouldn't be nice to warn the user?

**Comment by onderkalaci** [comment]
`citus_dist_stat_activity` will probably look like:
```SQL
CREATE VIEW citus_dist_stat_activity AS 
  SELECT global_pid,
         query,
         state
  FROM citus_stat_activity
  WHERE COALESCE(worker_query, FALSE) IS FALSE -- if NULL, it is a distributed command blocked on a DDL
    AND backend_type = 'client backend' ;
```

**Comment by onderkalaci** [comment]
> citus_dist_stat_activity will probably look like:

Well, probably not. 
```SQL
 AND backend_type = 'client backend' ;
```
When we don't have enough privileges as a user, backend_type is NULL.

So, thinking a little more on this

File: src/backend/distributed/sql/udfs/citus_stat_activity/latest.sql (hunk 6 above)
Comment (halilozanakgul) [reply src/backend/distributed/sql/udfs/citus_stat_activity/latest.sql]: I added a `CREATE VIEW` query to the PG upgrade function. So now it updates the view.

Comment (onderkalaci) [reply src/backend/distributed/sql/udfs/citus_stat_activity/latest.sql]: Cool, maybe quickly check with Nils if this sounds OK for him as well?

**Review by onderkalaci — COMMENTED** [review]
As we approach merging, can you please add a PR description?

I'll test this PR with https://github.com/citusdata/citus/pull/5748 then approve (the warning is optional, lets not spend too much time)

File: src/backend/distributed/sql/udfs/citus_finish_pg_upgrade/latest.sql [hunk 10]
```diff
@@ -137,6 +137,20 @@ BEGIN
...
+        SELECT * FROM run_command_on_all_nodes($$
+            SELECT coalesce(to_json(array_agg(csa_1n.*)), '[{}]'::JSON)
+            FROM (
+                SELECT global_pid, worker_query, (pg_stat_activity.*)::pg_stat_activity FROM
+                pg_stat_activity LEFT JOIN get_all_active_transactions() ON process_id = pid
+            ) AS csa_1n;
+        $$)
+        WHERE success = 't'
+    ) AS run_command_on_all_nodes
+    JOIN LATERAL json_populate_recordset(NULL::citus_stat_activity_one_node, run_command_on_all_nodes.result::JSON) as all_csa_1n ON TRUE;
```
[the remark this answers is not in the document]
Comment (halilozanakgul) [reply src/backend/distributed/sql/udfs/citus_finish_pg_upgrade/latest.sql:153]: No, it turned it is the view.
Also the tests pass this way.

File: src/backend/distributed/sql/udfs/citus_stat_activity/latest.sql [hunk 11]
```diff
@@ -0,0 +1,24 @@
+--DROP VIEW IF EXISTS pg_catalog.citus_stat_activity;
+--DROP TYPE IF EXISTS pg_catalog.citus_stat_activity_one_node;
+
+CREATE TYPE citus.citus_stat_activity_one_node AS (global_pid bigint, worker_query boolean, pg_stat_activity pg_stat_activity);
+
+ALTER TYPE citus.citus_stat_activity_one_node SET SCHEMA pg_catalog;
+GRANT ALL ON TYPE pg_catalog.citus_stat_activity_one_node TO PUBLIC;
+
+CREATE VIEW citus.citus_stat_activity AS
+SELECT all_csa_1n.global_pid, nodeid, all_csa_1n.worker_query, (all_csa_1n.pg_stat_activity).*
```
Comment (marcocitus) [inline src/backend/distributed/sql/udfs/citus_stat_activity/latest.sql:10]: would it make sense to call this `is_worker_query` to make it clear it's not a query?

File: src/backend/distributed/sql/udfs/citus_prepare_pg_upgrade/latest.sql [hunk 12]
```diff
@@ -67,6 +69,22 @@ BEGIN
        objects.colocationid
     FROM citus.pg_dist_object objects,
          pg_catalog.pg_identify_object_as_address(objects.classid, objects.objid, objects.objsubid) address;
+
+    SELECT count(*) > 0
```
Comment (onderkalaci) [inline src/backend/distributed/sql/udfs/citus_prepare_pg_upgrade/latest.sql:73]: minor: needs comment? What are we doing here? Why are we dropping citus_stat_activity?

File: src/backend/distributed/sql/udfs/run_command_on_all_nodes/latest.sql [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+	FROM citus_nodes;
+
+	SELECT count(*) > 0 FROM pg_dist_node
+	WHERE isactive = 't'
+	AND nodecluster = current_setting('citus.cluster_name')
+	AND groupid IN (SELECT groupid FROM pg_dist_local_group)
+	INTO current_node_is_in_metadata;
+
+	RETURN QUERY
+	SELECT pg_dist_node.nodeid, mrow.success, mrow.result
```
Comment (onderkalaci) [inline src/backend/distributed/sql/udfs/run_command_on_all_nodes/latest.sql:30]: as we discussed earlier, lets add  the current worker with `select current_setting('port');select current_setting('citus.local_hostname');`

 and avoid `EXECUTE command INTO ...` as that only works with SELECTs

File: src/backend/distributed/sql/udfs/citus_prepare_pg_upgrade/latest.sql (hunk 12 above)
Comment (onderkalaci) [inline src/backend/distributed/sql/udfs/citus_prepare_pg_upgrade/latest.sql:73]: I think the question is: When does `citus_depends_on_citus_stat_activity` ever becomes false?

File: src/backend/distributed/sql/udfs/run_command_on_all_nodes/latest.sql [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+		FROM master_run_on_worker(nodenames, ports, commands, parallel) mrow
+		JOIN pg_dist_node ON mrow.node_name = pg_dist_node.nodename AND mrow.node_port = pg_dist_node.nodeport;
+
+	IF NOT current_node_is_in_metadata THEN
```
[the remark this answers is not in the document]
Comment (halilozanakgul) [reply src/backend/distributed/sql/udfs/run_command_on_all_nodes/latest.sql:29]: I couldn't make it single line 🙃

File: src/backend/distributed/sql/udfs/citus_prepare_pg_upgrade/latest.sql (hunk 12 above)
Comment (halilozanakgul) [reply src/backend/distributed/sql/udfs/citus_prepare_pg_upgrade/latest.sql:73]: We call this function again to check if it is idempotent

Comment (onderkalaci) [reply src/backend/distributed/sql/udfs/citus_prepare_pg_upgrade/latest.sql:73]: let's add this as a comment as well, thanks

Comment (onderkalaci) [reply src/backend/distributed/sql/udfs/citus_prepare_pg_upgrade/latest.sql:73]: ah, already added

File: src/backend/distributed/sql/citus--10.2-4--11.0-1.sql [hunk 15]
```diff
@@ -80,3 +80,13 @@ BEGIN
       UPDATE pg_dist_node_metadata SET metadata=jsonb_set(metadata, '{partitioned_citus_table_exists_pre_11}', to_jsonb(partitioned_table_exists), true);
 END;
 $$;
+
+#include "udfs/run_command_on_all_nodes/11.0-1.sql"
+
+DROP VIEW IF EXISTS pg_catalog.citus_stat_activity CASCADE;
```
Comment (onderkalaci) [inline src/backend/distributed/sql/citus--10.2-4--11.0-1.sql:86]: These two seems unnecessary. How can we have these pre-11? If we are to follow this practice, we should do it for all types/views we create in upgrade scripts?
```SQL
DROP VIEW IF EXISTS pg_catalog.citus_stat_activity CASCADE;
DROP TYPE IF EXISTS pg_catalog.citus_stat_activity_one_node CASCADE;
```

**Comment by onderkalaci** [comment]
The tests fail because of https://github.com/citusdata/citus/pull/5749, lets get that first.

File: src/backend/distributed/sql/downgrades/citus--11.0-1--10.2-4.sql [hunk 16]
```diff
@@ -94,7 +94,7 @@ ALTER TABLE citus.pg_dist_object DROP COLUMN force_delegation;
 SET search_path = 'pg_catalog';
 
 
-DROP FUNCTION IF EXISTS get_all_active_transactions();
+DROP FUNCTION IF EXISTS get_all_active_transactions() CASCADE;
```
Comment (JelteF) [inline src/backend/distributed/sql/downgrades/citus--11.0-1--10.2-4.sql:97]: CASCADE seems scary. If a user somehow created a dependency between a table and this function, then we will inadvertently drop this table. Why do we need this?

File: src/backend/distributed/sql/udfs/run_command_on_all_nodes/latest.sql [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+	-- the coordinator is not added to the metadata.
+	-- We'll manually add current node to the lists to actually run on all nodes.
+	-- But when the coordinator is not added to metadata and this function
+	-- is called from a worker node, this will not be enough and we'll
+	-- not be able run on all nodes.
+	IF NOT current_node_is_in_metadata THEN
+		SELECT
+		array_append(nodenames, current_setting('citus.local_hostname')),
+		array_append(ports, current_setting('port')::int),
+		array_append(commands, command)
```
Comment (onderkalaci) [inline src/backend/distributed/sql/udfs/run_command_on_all_nodes/latest.sql:45]: II cannot follow why we append the command again. Are we running the commands multiple times?

File: src/backend/distributed/sql/udfs/citus_finish_pg_upgrade/latest.sql [hunk 18]
```diff
@@ -144,6 +145,22 @@ BEGIN
         pg_catalog.pg_get_object_address(naming.type, naming.object_names, naming.object_args) address;
 
     DROP TABLE public.pg_dist_object;
+
+    -- we recreate citus_stat_activity after PG updates so it reflects the changes in pg_stat_activity
+    #include "../../udfs/citus_stat_activity/11.0-1.sql"
+
+    SELECT count(*) > 0
```
Comment (onderkalaci) [inline src/backend/distributed/sql/udfs/citus_finish_pg_upgrade/latest.sql:152]: this is to make the operation idempotent

File: src/backend/distributed/sql/udfs/citus_prepare_pg_upgrade/latest.sql [hunk 19]
```diff
@@ -67,6 +69,25 @@ BEGIN
...
+    FROM pg_depend
```
Comment (onderkalaci) [inline src/backend/distributed/sql/udfs/citus_prepare_pg_upgrade/latest.sql:74]: this is to make the operation idempotent

File: src/backend/distributed/sql/udfs/citus_prepare_pg_upgrade/latest.sql [hunk 20]
```diff
@@ -67,6 +69,25 @@ BEGIN
...
+    -- drop citus_stat_activity to recreate in citus_prepare_pg_upgrade
+    -- so it has all the columns in pg_stat_activity, even if the upgrade changes them
+    -- also we need to check if it exists to make this function idempotent
+    IF citus_depends_on_citus_stat_activity THEN
+        ALTER EXTENSION citus DROP VIEW citus_stat_activity;
+        ALTER EXTENSION citus DROP TYPE citus_stat_activity_one_node;
+    END IF;
+
+    DROP VIEW IF EXISTS pg_catalog.citus_stat_activity;
+    DROP TYPE IF EXISTS pg_catalog.citus_stat_activity_one_node;
```
Comment (JelteF) [inline src/backend/distributed/sql/udfs/citus_prepare_pg_upgrade/latest.sql:90]: Instead of dropping the VIEW in prepare_pg_upgrade and recreating it finish_pg_upgrade, I think an easier  way is to create the view using `CREATE OR REPLACE`. Then there's no need to do all this pg_depend stuff (I think).

File: src/backend/distributed/sql/udfs/citus_stat_activity/latest.sql [hunk 21]
```diff
@@ -0,0 +1,25 @@
+SET citus.enable_metadata_sync TO OFF;
+
+CREATE TYPE citus.citus_stat_activity_one_node AS (global_pid bigint, is_worker_query boolean, pg_stat_activity pg_stat_activity);
+
+ALTER TYPE citus.citus_stat_activity_one_node SET SCHEMA pg_catalog;
+GRANT ALL ON TYPE pg_catalog.citus_stat_activity_one_node TO PUBLIC;
```
Comment (JelteF) [inline src/backend/distributed/sql/udfs/citus_stat_activity/latest.sql:6]: I think this is the same currently, as USAGE is the only permission on a type (afaict), but making that explicit seems good from a clarity and security perspective.
```suggestion
GRANT USAGE ON TYPE pg_catalog.citus_stat_activity_one_node TO PUBLIC;
```

File: src/backend/distributed/sql/udfs/citus_prepare_pg_upgrade/latest.sql (hunk 20 above)
Comment (onderkalaci) [reply src/backend/distributed/sql/udfs/citus_prepare_pg_upgrade/latest.sql:90]: Hmm, I think we tried that didn't work for some reason. But let me retry that actually.

Comment (onderkalaci) [reply src/backend/distributed/sql/udfs/citus_prepare_pg_upgrade/latest.sql:90]: I think that was because we also need a type `citus_stat_activity_one_node` which we have to drop (because there is not create_or_replace_type in Postgres. Maybe we can try `worker_create_or_replace_object` hack

File: src/backend/distributed/sql/udfs/run_command_on_all_nodes/latest.sql (hunk 17 above)
Comment (halilozanakgul) [reply src/backend/distributed/sql/udfs/run_command_on_all_nodes/latest.sql:45]: `master_run_on_worker` expects a separate query for each node

File: src/backend/distributed/sql/udfs/citus_prepare_pg_upgrade/latest.sql (hunk 20 above)
Comment (halilozanakgul) [reply src/backend/distributed/sql/udfs/citus_prepare_pg_upgrade/latest.sql]: I removed this but for the record, it was in case a column was dropped from `pg_stat_activity`. If you don't drop the table before upgrade, and a column is dropped in upgrade it failed.

File: src/backend/distributed/sql/udfs/citus_stat_activity/latest.sql (hunk 21 above)
Comment (halilozanakgul) [reply src/backend/distributed/sql/udfs/citus_stat_activity/latest.sql]: Great suggestion but I dropped this type too

File: src/backend/distributed/sql/udfs/citus_stat_activity/latest.sql [hunk 22]
```diff
@@ -0,0 +1,42 @@
+-- citus_stat_activity combines the pg_stat_activity views from all nodes and adds global_pid, nodeid and is_worker_query columns.
```
Comment (onderkalaci) [inline src/backend/distributed/sql/udfs/citus_stat_activity/latest.sql:1]: There is still a possibility that Postgres drop some column from pg_stat_activity, where we have to drop.

In that case, we cannot do `CREATE OR REPLACE ciitus_stat_activity` because of this: https://www.postgresql.org/message-id/flat/BYAPR16MB2789E8E00904B6ED17FE1955BE9B0%40BYAPR16MB2789.namprd16.prod.outlook.com

In that case, we have to do `DROP VIEW citus_stat_activity CASCADE`. Or, we can keep that column until a major Citus release and return NULL.

I think that is OK

File: src/backend/distributed/sql/udfs/citus_stat_activity/latest.sql [hunk 23]
```diff
@@ -0,0 +1,42 @@
...
+
+CREATE OR REPLACE FUNCTION pg_catalog.citus_stat_activity(OUT global_pid bigint, OUT nodeid int, OUT is_worker_query boolean, OUT datid oid, OUT datname name, OUT pid integer,
+                                                          OUT leader_pid integer, OUT usesysid oid, OUT usename name, OUT application_name text, OUT client_addr inet, OUT client_hostname text,
+                                                          OUT client_port integer, OUT backend_start timestamp with time zone, OUT xact_start timestamp with time zone,
+                                                          OUT query_start timestamp with time zone, OUT state_change timestamp with time zone, OUT wait_event_type text, OUT wait_event text,
+                                                          OUT state text, OUT backend_xid xid, OUT backend_xmin xid, OUT query_id bigint, OUT query text, OUT backend_type text)
+    RETURNS SETOF record
+    LANGUAGE plpgsql
+    AS $function$
+BEGIN
```
Comment (onderkalaci) [inline src/backend/distributed/sql/udfs/citus_stat_activity/latest.sql:16]: For future reference: `jsonb_to_recordset` does what we like. If a column (say query_id) isn't available in the jsonb, the return recordset contains `NULL`.

**Review by onderkalaci — APPROVED** [review]
Thanks for the smart solution!

If we don't have any, can you consider adding a (consistent output) single test where a non=superuser accesses to `citus_stat_activity`?

A simple query like `select count(*) from citus_stat_activity WHERE global_pid = citus_backend_gpid() and NOT is_worker_query;` should always  return 1?

(Maybe we merge this one and I can add the test to #5755)
